lib/ngx/re.lua

```diff
@@ -0,0 +1,163 @@
+-- I hereby assign copyright in this code to the lua-resty-core project,
+-- to be licensed under the same terms as the rest of the code.
+
+local ffi = require 'ffi'
+local bit = require "bit"
+local base = require "resty.core.base"
+local core_regex = require "resty.core.regex"
+
+
+local C = ffi.C
+local sub = string.sub
+local band = bit.band
+local new_tab = base.new_tab
+local tostring = tostring
+local re_match_compile = core_regex.re_match_compile
+local destroy_compiled_regex = core_regex.destroy_compiled_regex
+
+
+local PCRE_ERROR_NOMATCH = -1
+
+
+local FLAG_DFA           = 0x02
+
+
+ffi.cdef [[
+    int ngx_http_lua_ffi_exec_regex(ngx_http_lua_regex_t *re, int flags,
+        const unsigned char *s, size_t len, int pos);
+]]
+
+
+local _M = { version = base.version }
+
+
+local function re_split_helper(subj, regex, opts, ctx)
+    local compiled, compile_once, flags = re_match_compile(regex, opts)
```

Comment:
This is unnecessarily inefficient to compile the same regex multiple times (or query the compiled regex cache when `o` opt is specified) in a single `split` call. We should cache the results of `re_match_compile` outside the `while` loop in the `split` function.